Sources/TensorFlow/Operators/Math.swift

```diff
@@ -1397,11 +1401,12 @@ internal func _vjpMinMaxHelper<T: TensorFlowFloatingPoint>(
     _ x: Tensor<T>,
     _ y: Tensor<T>,
     originalValue: Tensor<T>,
-    seed: Tensor<T>
+    seed: Tensor<T>,
+    comparisonOperation: (Tensor<T>, Tensor<T>) -> Tensor<Bool>
 ) -> (Tensor<T>, Tensor<T>) {
-    let denominator = 1 + Tensor<T>(x .== y)
-    let lhsGrad = seed * Tensor<T>(x .== originalValue) / denominator
-    let rhsGrad = seed * Tensor<T>(y .== originalValue) / denominator
+    let mask = Tensor<T>(comparisonOperation(x, y))
+    let lhsGrad = seed * mask
+    let rhsGrad = seed - lhsGrad
```

Comment:
It may be better to use `seed * (1 - mask)` here from a numerical precision point of view.

Reply:
Could we please try to match TensorFlow gradient implementations exactly (see [here](https://github.com/tensorflow/swift-apis/pull/480#discussion_r317415394))?

I believe that involves using `let rhsGrad = seed - lhsGrad`: Python TensorFlow reference implementation [here](https://github.com/tensorflow/tensorflow/blob/dac733509fbe366804656a71f15ba0712bb5e69d/tensorflow/core/ops/math_grad.cc#L614).

Reply:
@dan-zheng I believe the code you're referring to is not that of the Python API gradients, but rather the C++ API gradients. The Python API gradients are defined [here](https://github.com/tensorflow/tensorflow/blob/752556174a1dcf68693784a8c78f07ba818e230f/tensorflow/python/ops/math_grad.py#L213). I initially implemented the gradients of the min and max to match that implementation, before this PR. @rxwei which implementation were you referring to in #479?

In general, it seems that there are multiple versions of these gradients in the various TF APIs. All are correct given that min/max have a non-differentiable point. All we care about is for a subgradient and all these implementations provide valid subgradients and should all pretty much result in the same behavior as it's highly unlikely a value would equal the non-differentiable point (unless intentionally done so).

I understand that we're using the Python API to generally verify gradients so we should probably stick to what's done there so differences don't accumulate and make it. hard to verify other gradients. Could you please confirm that [this](https://github.com/tensorflow/tensorflow/blob/752556174a1dcf68693784a8c78f07ba818e230f/tensorflow/python/ops/math_grad.py#L213) is the Python API implementation? If so, @rxwei could you please point to the issue related to #479?

Reply:
I left [a comment](https://github.com/tensorflow/swift-apis/issues/479#issuecomment-524669755) on #479 describing the exact gradient calculation mismatch for `tf.minimum`/`tf.maximum`.

> @dan-zheng I believe the code you're referring to is not that of the Python API gradients, but rather the C++ API gradients. The Python API gradients are defined [here](https://github.com/tensorflow/tensorflow/blob/752556174a1dcf68693784a8c78f07ba818e230f/tensorflow/python/ops/math_grad.py#L213).

Thanks for pointing out the distinction between Python and C++ gradient implementations! Do you happen to know when each implementation is used?

(It seems likely to me that gradients registered in Python via `@ops.RegisterGradient` are used to compute gradients in Python - that makes me wonder when the C++ gradient implementations are used.)

---

You linked to `_MinOrMaxGrad`, but I think we actually want to match [`_MaximumMinimumGrad`](https://github.com/tensorflow/tensorflow/blob/752556174a1dcf68693784a8c78f07ba818e230f/tensorflow/python/ops/math_grad.py#L1417) (which seems similar but perhaps subtly different from the current Swift implementation in this PR). The bottom line is the gradient calculation mismatch should be fixed 🙂

Reply:
For what it's worth, I could not produce gradients equivalent to the Python TF 2.0-Beta implementation when using the previous implementation.

Reply:
@dan-zheng you're right sorry. In that case, we should switch to `seed * (1 - mask)` to match that behavior. Thanks for helping clarify this! Regarding the Python vs C++ gradients, I had asked that same question a couple years back and I think the answer was that C++ gradients are only used by the C++ API for TensorFlow (i.e., not the Python API or the runtime). The Python API gradients are all registered in Python and the only way they can have C++ implementations is through custom ops and corresponding kernels. @alextp may be able to verify this.